Move dataset op names into name_utils namespace

feihugis opened the pull request:
This PR aims to move all dataset op names into `tensorflow::data::name_utils` namespace and replace all uses of string literals for dataset names.

cc: @jsimsa

feihugis commented on tensorflow/core/kernels/data/map_dataset_op.cc:
```diff
@@ -77,8 +80,8 @@ class MapDatasetOp : public UnaryDatasetOpKernel {
-      return absl::make_unique<Iterator>(
-          Iterator::Params{this, strings::StrCat(prefix, "::Map")});
+      return absl::make_unique<Iterator>(Iterator::Params{
+          this, strings::StrCat(prefix, "::", kMapDatasetOpName)});
```
FYI, this change will make the prefix different with before (e.g. `Iterator::Map` --> `Iterator::MapDataset`).

jsimsa replied:
This should be avoided because I am aware of tools that parse and analyze the labels (which show up in CPU profile traces).

jsimsa replied:
Alright, it is used in exactly one place (in dataset.cc) and I think we can replace it with a string based on `dataset_type_` once your PR is submitted.

jsimsa commented on tensorflow/core/kernels/data/name_utils.cc:
```diff
@@ -0,0 +1,28 @@
+namespace data {
+namespace name_utils {
+
+ABSL_CONST_INIT const char kRangeDatasetOpName[] = "RangeDataset";
```
Instead of doing this, you should:

1) Create a `DatasetType` enum
2) Create a `ToString()` method that takes an enum and returns the "base" name. The base name should match the string used in the iterator prefix.
3) Create a `DatasetDebugString()` method that takes an enum (and variadic arguments list of arguments) returns the equivalent `DebugString` version of the name "FooDatasetOp(arg1, arg2, ...)::Dataset"
4) Create a `IteratorPrefix()` method that takes an enum and a prefix for downstream iterator and returns the prefix "Iterator::Map::Range"

Individual dataset would then define simply declare their type at construction time and do not do any string manipulation themselves.

jsimsa reviewed (changes requested):
Thanks Fei, left a couple of comments.

feihugis committed:
Move Range&Map dataset op names into name_utils namespace

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op.cc:
```diff
@@ -27,6 +28,8 @@ namespace {
 // See documentation in ../../ops/dataset_ops.cc for a high-level
 // description of the following op.
 
+using namespace tensorflow::data::name_utils;
```
remove this and instead use `name_utils::IteratorPrefix` and `name_utils::DatasetDebugString` ... importing entire namespace "pollutes" the context with all of its symbols

feihugis replied:
Done

feihugis commented:
@jsimsa Thanks for your comments. I rewrite this PR according to your comments. Could you please have a look at the changes (https://github.com/tensorflow/tensorflow/pull/28531/commits/e81ba1947142deb0fb4424cbc30d7f11a8245763)?

jsimsa commented on tensorflow/core/kernels/data/name_utils.cc:
```diff
@@ -0,0 +1,68 @@
+
+string ToString(const DatasetType& dataset_type) {
+  switch (dataset_type) {
+    case RANGE_DATASET:
```
no need for the `_DATASET` suffix, it is implied by the enum type

feihugis replied:
Done

jsimsa commented on tensorflow/core/kernels/data/name_utils.h:
```diff
@@ -0,0 +1,59 @@
+namespace name_utils {
+using namespace tensorflow::strings;
+
+enum DatasetType {
```
order alphabetically and drop the `_DATASET` suffix

feihugis replied:
Done

jsimsa commented on tensorflow/core/kernels/data/name_utils.h:
```diff
@@ -0,0 +1,59 @@
+// Returns the dataset op name.
+string OpName(const DatasetType& dataset_type);
+
+string DatasetDebugString(const DatasetType& dataset_type,
```
do we need this if we have the other one?

feihugis replied:
Tried to combine these two functions into one as below:
```C++
template <typename... Args>
string DatasetDebugString(const DatasetType& dataset_type,
                          const Args&... args) {
  std::initializer_list<StringPiece> arg_list{
      static_cast<const strings::AlphaNum&>(args).Piece()...};

  if (arg_list.size() == 0) {
    return strings::StrCat(OpName(dataset_type), "Op::Dataset");
  }

  string debug_str;
  strings::StrAppend(&debug_str, OpName(dataset_type), "Op(");
  auto iter = arg_list.begin();
  while (iter != arg_list.end() - 1) {
    strings::StrAppend(&debug_str, *iter, ", ");
    ++iter;
  }
  strings::StrAppend(&debug_str, *iter, ")::Dataset");
  return debug_str;
}
```
However, the return output is not expected. For example, for `RangeDataset`, the output changes to `RangeDatasetOp(0, 10, �)::Dataset` rather than `RangeDatasetOp(0, 10, 3)::Dataset`.

@jsimsa Do you have any suggestion?

jsimsa commented on tensorflow/core/kernels/data/name_utils.h:
```diff
@@ -0,0 +1,59 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
For all the public methods in this API, provide an example in the documentation (given inputs "foo", method returns "bar").

feihugis replied:
Done

feihugis commented on tensorflow/core/kernels/data/map_dataset_op.cc:
```diff
@@ -88,7 +91,9 @@ class MapDatasetOp : public UnaryDatasetOpKernel {
 
-    string DebugString() const override { return "MapDatasetOp::Dataset"; }
+    string DebugString() const override {
+      return DatasetDebugString(MAP_DATASET);
```
if we add `dataset_type_` member for each dataset, will there be some overlap between `dataset_type_` and [type_string_](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/dataset.h#L713)?

jsimsa replied:
As far as I can tell, `type_string_` is not used for anything (other than in your unit tests), so it might be okay to get rid of it. Let me try creating a change that does that.

feihugis replied:
Feel free to let me know if there is anything I can help.

jsimsa reviewed (changes requested):
Thanks Fei, looks much better. Did another pass and left a couple more comments.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -36,6 +37,8 @@ limitations under the License.
 namespace tensorflow {
 namespace data {
 
+using namespace tensorflow::data::name_utils;
```
remove this

feihugis replied:
Done

feihugis committed:
Address the commments
- Avoid the importing of entire namespace, which "pollutes" the context with all of its symbols
- Drop _DATASET suffix
- Examples for the public API

feihugis commented:
@jsimsa The comments are addressed here (https://github.com/tensorflow/tensorflow/pull/28531/commits/885715e4deeafb9ab20afb959e7f98e561b4dd07), except [this one](https://github.com/tensorflow/tensorflow/pull/28531#discussion_r282693147). More details can be found at the comments.

jsimsa commented:
@feihugis thank you for the making the changes. After looking at the CL, I realized that it is probably not a good idea to expect all dataset types to be added to the `enum` in `name_utils.h`. After all, there are dataset implementation that do not exist in our code base.

So instead of having an enum, I suggest to make the base op name be a static member of the op kernel. We would still keep the `name_utils::XXX` method that take a base string (instead of enum) and return the different versions of it.

To make the base op name accessible from unit tests, we will want to create .h files for ops that do not have them that define the externally visible op interface. Take a look at `prefetch_dataset_op.{h, cc}` for a minimal example of how to do this.

I probably makes sense to do this incrementally for existing ops, starting with ops that already have unit tests. Let me know if you have questions. Thanks!

feihugis commented:
@jsimsa I agree with you and have one question in below.

> After looking at the CL, I realized that it is probably not a good idea to expect all dataset types to be added to the `enum` in `name_utils.h`. After all, there are dataset implementation that do not exist in our code base.
> 
Yes, it is a good point, e.g. the datasets in tf-io.

> So instead of having an enum, I suggest to make the base op name be a static member of the op kernel. We would still keep the `name_utils::XXX` method that take a base string (instead of enum) and return the different versions of it.
> 
> To make the base op name accessible from unit tests, we will want to create .h files for ops that do not have them that define the externally visible op interface. Take a look at `prefetch_dataset_op.{h, cc}` for a minimal example of how to do this.
>
As we already have `type_string_` member in `DatasetBase`, could we use `type_string_` as the input for `name_utils::XXX` (we can also add `name_utils::Basename()` to return the base op name) so that we do not need to create a new static member or .h files.

jsimsa commented:
The problem with type_string_ is that it is derived from the op name and multiple ops can use the same kernel (e.g. BatchDataset and BatchDatasetV2). In general we do not want to distinguish between different ops executing the same kernel but there are some exceptions to that rule (e.g. ParallelInterleave and ParallelInterleaveV2 which have different op kernels).

If you do not want to go through the trouble of creating a new static member (`kDatasetTypeName`) and header files, that's fine. I am okay with not doing this refactoring.

feihugis commented:
> The problem with `type_string_` is that it is derived from the op name and multiple ops can use the same kernel (e.g. BatchDataset and BatchDatasetV2). In general we do not want to distinguish between different ops executing the same kernel but there are some exceptions to that rule (e.g. ParallelInterleave and ParallelInterleaveV2 which have different op kernels).
> 
> If you do not want to go through the trouble of creating a new static member (`kDatasetTypeName`) and header files, that's fine. I am okay with not doing this refactoring.

@jsimsa Thanks for your explanation! I understand now. I do want to try the refactoring with header files. Will submit a commit as an example for your review.

feihugis committed:
Refactor {Range,Map,Prefectch}Dataset using name_utils

feihugis commented:
@jsimsa The {Range, Map, Prefetch} datasets have been refactored with the new static member (`kDatasetTypeName`), header files, and `name_utils`. Could you please have a look at the changes (https://github.com/tensorflow/tensorflow/pull/28531/commits/d2bd4afee7fb31415704f24b0766ab26d0d2a87f) when you have time?

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -28,6 +28,7 @@ limitations under the License.
 #include "tensorflow/core/framework/variant_tensor_data.h"
 #include "tensorflow/core/kernels/data/dataset_utils.h"
 #include "tensorflow/core/kernels/data/iterator_ops.h"
+#include "tensorflow/core/kernels/data/name_utils.h"
```
not needed?

feihugis replied:
Yes, will remove it.

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op.cc:
```diff
@@ -12,219 +12,194 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+                        output_shapes_, preserve_cardinality_);
+}
+
 REGISTER_KERNEL_BUILDER(Name("MapDataset").Device(DEVICE_CPU), MapDatasetOp);
```
put the kernel registration into anonymous namespace ... applies to other .cc files as well

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op.cc:
```diff
@@ -12,219 +12,194 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // description of the following op.
 
-class MapDatasetOp : public UnaryDatasetOpKernel {
+constexpr char MapDatasetOp::kDatasetType[];
```
what do we need this for?

feihugis replied:
@jsimsa Without this, there will be some errors when running the unit test:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //tensorflow/core/kernels/data:map_dataset_op_test
-----------------------------------------------------------------------------
dyld: Symbol not found: __ZN10tensorflow4data12MapDatasetOp12kDatasetTypeE
  Referenced from: /private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/core/kernels/data/map_dataset_op_test.runfiles/org_tensorflow/tensorflow/core/kernels/data/map_dataset_op_test
  Expected in: flat namespace
 in /private/var/tmp/_bazel_fei/5dc7b372a7c45427ff30500d3e22fe26/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/core/kernels/data/map_dataset_op_test.runfiles/org_tensorflow/tensorflow/core/kernels/data/map_dataset_op_test
external/bazel_tools/tools/test/test-setup.sh: line 310: 19316 Abort trap: 6           "${TEST_PATH}" "$@" 2>&1
```

jsimsa commented on tensorflow/core/kernels/data/map_dataset_op_test.cc:
```diff
@@ -13,6 +13,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#include "tensorflow/core/kernels/data/map_dataset_op.h"
```
this import seems out of order

feihugis replied:
clang-format suggest me to move it to the top. Should I organize them alphabetically?

feihugis replied:
As this [example](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/attr_value_util_test.cc#L16) shows, this header file may be put at the top.

jsimsa commented on tensorflow/core/kernels/data/name_utils.cc:
```diff
@@ -0,0 +1,61 @@
+  } else if (dataset_type == "Range") {
+    return "RangeDataset";
+  }
+  LOG(FATAL) << "Unknown dataset type " << dataset_type << std::endl;
```
FATAL will abort the execution ... this should be WARNING

jsimsa commented on tensorflow/core/kernels/data/name_utils.cc:
```diff
@@ -0,0 +1,61 @@
+    return "RangeDataset";
+  }
+  LOG(FATAL) << "Unknown dataset type " << dataset_type << std::endl;
+  return "Unknown dataset type";
```
UnknownDataset

jsimsa commented on tensorflow/core/kernels/data/name_utils.cc:
```diff
@@ -0,0 +1,61 @@
+  return "Unknown dataset type";
+}
+
+namespace internal {
```
why the internal namespace?

feihugis replied:
In the beginning, I thought this function should not be public, so move it to the internal namespace. Will remove it.

jsimsa commented on tensorflow/core/kernels/data/name_utils.cc:
```diff
@@ -0,0 +1,61 @@
+}  // namespace internal
+
+string IteratorPrefix(const string& dataset_type, const string& prefix) {
+  return strings::StrCat(prefix, "::", dataset_type);
```
create a named constant `kPrefixDelimiter` for `::`

jsimsa reviewed (changes requested):
@feihugis thank you for the changes ... I left some more comments.

Note that I will be traveling next week and will not be available to review your changes.

feihugis committed:
Address the comments

feihugis commented:
@jsimsa This commit (https://github.com/tensorflow/tensorflow/pull/28531/commits/ec7f39c4e8665e98e41e6340737327ffb69389d9) is submitted to address your comments, except [this one](https://github.com/tensorflow/tensorflow/pull/28531#discussion_r283102963). 

Thanks for letting me know your travel! Have a safe and enjoyable trip!

jsimsa reviewed (changes requested):
Thanks @feihugis left a couple of comments all related to making all input and attribute names named constants (similar to `kDatasetType`). Other than that, this looks good to me so you can remove the WIP tag.

feihugis commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -92,7 +94,8 @@ class DatasetOpsTestBase : public ::testing::Test {
     NodeDef node_def = test::function::NDef(
-        node_name, "RangeDataset", {"start", "stop", "step"},
+        node_name, name_utils::OpName(RangeDatasetOp::kDatasetType),
+        {"start", "stop", "step"},
```
Yes, will do.

feihugis committed:
Make all input and attribute names named constants

feihugis commented:
@jsimsa The dataset input and attribute names are changed to be named constants, and also make a little refactoring for `RangeDatasetOpTest` and `MapDatasetOpTest` via (https://github.com/tensorflow/tensorflow/pull/28531/commits/60f6c35f6a5dd9405d11c447222178684a766e33). Could you please have a look at the changes when you have time?

jsimsa commented on tensorflow/core/kernels/data/name_utils.cc:
```diff
@@ -0,0 +1,59 @@
+string DatasetDebugString(const string& dataset_type,
+                          std::initializer_list<StringPiece> args) {
+  if (args.size() == 0) {
+    return strings::StrCat(dataset_type, "Op::Dataset");
```
use `kDelimiter` here

jsimsa commented on tensorflow/core/kernels/data/name_utils.cc:
```diff
@@ -0,0 +1,59 @@
+namespace data {
+namespace name_utils {
+
+ABSL_CONST_INIT const char kPrefixDelimiter[] = "::";
```
rename to `kDelimiter`

jsimsa commented on tensorflow/core/kernels/data/name_utils.cc:
```diff
@@ -0,0 +1,59 @@
+    strings::StrAppend(&debug_str, *iter, ", ");
+    ++iter;
+  }
+  strings::StrAppend(&debug_str, *iter, ")::Dataset");
```
ditto

jsimsa reviewed (commented):
@feihugis left a couple of minor comments, otherwise this looks great ... thanks!

feihugis committed:
Replace the delimiter in name_utlis

feihugis commented:
@jsimsa Good suggestion. Change them vis this commit (https://github.com/tensorflow/tensorflow/pull/28531/commits/91b412bd79f2d521b7b3a3c267c1ca568c83c843).

feihugis commented:
@jsimsa The internal checks failed. Could you help paste the log here?

jsimsa commented:
```
third_party/tensorflow/core/kernels/data/prefetch_dataset_op_test.cc:16:10: error: module //third_party/tensorflow/core/kernels/data:prefetch_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/kernels/data/name_utils.h'
see http://go/cpp-features#layering_check; to fix run:
build_cleaner //third_party/tensorflow/core/kernels/data:prefetch_dataset_op_test
#include "third_party/tensorflow/core/kernels/data/name_utils.h"
```

feihugis committed:
Remove the redundant import

feihugis commented:
This import `#include "tensorflow/core/kernels/data/name_utils.h` is not needed in `prefetch_dataset_op_test.cc` as it has been imported in `DatasetOpsTestBase`, so it is removed via this commit (https://github.com/tensorflow/tensorflow/pull/28531/commits/0928e53708715d11da6b425be20ccbe1fadc1770). @jsimsa Could you please have a look at the change?

feihugis commented:
@jsimsa Thanks for your help on this PR! I will start to refactor the rest dataset ops. Do you think which way is better to submit the PRs: 1) one PR for one dataset op; 2) one PR for several dataset ops?

jsimsa commented:
You can submit multiple of them in one PR, but use separate commits for separate dataset ops.

feihugis commented:
> You can submit multiple of them in one PR, but use separate commits for separate dataset ops.

Yeah, will do.
